framework/src/base/MeshGeneratorSystem.C

```diff
@@ -0,0 +1,494 @@
+  }
+
+  // ...and sort them
+  try
+  {
+    _ordered_mesh_generators = resolver.getSortedValuesSets();
+  }
+  // It is _quite_ hard to trigger this to test it. I've tried to no avail.
+  // Now that we...
+  // - check and sort up front
+  // - know if dependencies exist at the time of requesing them
+  // - require that sub generators depend only on other sub generators in an object's
+  //   tree + input dependencies that we explicitly declare
+  // I don't think it's possible. But we'll leave it here anyway and it
+  // definitely will not be covered
+  catch (CyclicDependencyException<MeshGenerator *, MeshGenerator::Comparator> & e)
+  {
+    const auto & cycle = e.getCyclicDependencies();
+    std::vector<std::string> names(cycle.size());
+    for (const auto i : index_range(cycle))
+      names[i] = cycle[i]->name();
+
+    mooseError("Cyclic dependencies detected in mesh generation: ",
+               MooseUtils::join(names, " <- "));
+  }
+
+  const auto & final_generators = _ordered_mesh_generators.back();
+
+  // We haven't forced a final generator yet
+  if (_final_generator_name.empty())
+  {
+    // See if we have multiple independent trees of generators
+    const auto ancestor_list = resolver.getAncestors(final_generators.back());
+    if (ancestor_list.size() != resolver.size() && _final_generator_name.empty())
+    {
+      // Need to remove duplicates and possibly perform a difference so we'll import out list
+      // into a set for these operations.
+      std::set<MeshGenerator *, MeshGenerator::Comparator> ancestors(ancestor_list.begin(),
+                                                                     ancestor_list.end());
+      // Get all of the items from the resolver so we can compare against the tree from the
+      // final generator we just pulled.
+      const auto & allValues = resolver.getSortedValues();
+      decltype(ancestors) all(allValues.begin(), allValues.end());
+
+      decltype(ancestors) ind_tree;
+      std::set_difference(all.begin(),
+                          all.end(),
+                          ancestors.begin(),
+                          ancestors.end(),
+                          std::inserter(ind_tree, ind_tree.end()));
+
+      std::ostringstream oss;
+      oss << "Your MeshGenerator tree contains multiple possible generator outputs :\n\""
+          << final_generators.back()->name()
+          << " and one or more of the following from an independent set: \"";
+      bool first = true;
+      for (const auto & gen : ind_tree)
+      {
+        if (!first)
+          oss << ", ";
+        else
+          first = false;
+
+        oss << gen->name();
+      }
+      oss << "\"\n\nThis may be due to a missing dependency or may be intentional. Please "
+             "select the final MeshGenerator in\nthe [Mesh] block with the \"final_generator\" "
+             "parameter or add additional dependencies to remove the ambiguity.";
```

Comment:
> The final mesh generator is a red herring 99% of the time
> Let s have it as the last option

Not sure what you're saying here. This is not new code.

> Cant the user reach this by having multiple final generators by accident? This would not be an issue with dependencies.

Still not sure what you're trying to say here. A user can absolutely hit this by accident, and they could be missing dependencies (mesh generators) that cause a tree to have multiple independent endpoints.

Reply:
reword should be clear to users now